networkgen/mapgen.py

```diff
@@ -0,0 +1,447 @@
+        for edge in edge_list:
+            if r1 == edge[0] or r1 == edge[1] or r2 == edge[0] or r2 == edge[1]:
+                return True # Shortcut comparing: it's already True
+        return False
+
+    def not_ingraph(self, node_list, candidate_edge):
+        r1, r2 = candidate_edge.split()[0], candidate_edge.split()[1]
+        return r1 not in node_list or r2 not in node_list
+
+    def outer_nodes(self, G):
+        node_list = []
+        for node in G.nodes:
+            if len(G.edges(node)) == 1:
+                node_list.append(node)
+        return node_list
+
+    def make_map(self):
+        for charge, lig in self.lig_dict.items():
+            H = nx.Graph()
+            if len(lig['Name']) == 1:  # In case one ligand is found alone in a charge group
+                ligcol = self.sim_dfs[lig['Name']].sort_values(by=[lig['Name'][0]]) #complete similarity matrix
+                H.add_edge(lig['Name'][0], ligcol.index[1])
+                H.add_edge(lig['Name'][0], ligcol.index[2])
+                lig['Graph'] = H
+                break
+
+            # 1. Make SPT
+            incomplete = True
+            while incomplete:
+                for pert, score in lig['pairs_dict'].items():
+                    if len(H.nodes) == len(lig['Name']):
+                        incomplete = False
+                        break
+                    l1, l2 = pert.split()[0], pert.split()[1]
+                    if H.has_edge(l1, l2) or H.has_edge(l2, l1):
+                        continue
+                    if len(H.nodes) == 0 or self.intersection(H.edges, pert) and self.not_ingraph(H.nodes, pert):
+                        H.add_edge(l1, l2, weight=score)
+                        break
+
+            # 2. Close Cycles
+            while len(self.outer_nodes(H)) != 0:
+                for pert, score in lig['pairs_dict'].items():
+                    l1, l2 = pert.split()[0], pert.split()[1]
+                    if l1 in self.outer_nodes(H) or l2 in self.outer_nodes(H):
+                        if (l1, l2) not in H.edges or (l2, l1) not in H.edges:
+                            H.add_edge(l1, l2, weight=score)
+                            break
+
+            # 3. Add influence edges
+            eig_cent = nx.eigenvector_centrality(H, max_iter=1000)
+            eig_cent = {k: v for k, v in sorted(eig_cent.items(), key=lambda item: item[1], reverse=True)}
+            per_nodes = [k for k,v in nx.eigenvector_centrality(H, max_iter=1000).items() if v < 0.01]
+            per_len = len(per_nodes)
+            while per_len > 1:
+                for pert, score in lig['pairs_dict'].items():
+                    l1, l2 = pert.split()[0], pert.split()[1]
+                    if l1 in per_nodes and l2 not in per_nodes or l1 not in per_nodes and l2 in per_nodes:
+                        if (l1, l2) not in H.edges or (l2, l1) not in H.edges and intersection(H.edges, pert):
```

Comment:
intersection is not defined, where is it coming from?

Reply:
intersection is a method of the same MapGen class. self.intersection.